### DISCUSSION TITLE
`extends` seems to overwrite local config

### DISCUSSION THREAD

**Discussion by fdw** [renovatebot/renovate#27767] [Request Help] [answered]
### What would you like help with?

I would like help with my configuration

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

BitBucket, `37.230.0`

### Please tell us more about your question or problem

We're using only a global config; the config in the repo only sets the schedule.

In the global config, we have:
```
module.exports = {
  <snip>
  "extends": ["group:monorepos", "group:recommended"],
  "packageRules: [
    {
      "matchDatasources": ["npm"],
      "matchPackagePatterns": ["^@reduxjs/", "react-redux"],
      "groupName": "Reduxjs"
    },
    <some other, similar groupings>
    {
      "matchUpdateTypes": ["minor", "patch"],
      "matchPackagePatterns": ["*"],
      "groupName": "Non-major"
    }
  ]
}
```

I would expect that all `minor` and `patch` level updates are now grouped into one PR. However, we still see other PRs with only minor/patch upgrades. They all seem to belong to some monorepos or other groups (f.e. `dotnet` or `Refit`).
How can we configure Renovate to group *all* `minor` and `patch`, regardless of other `packageRules`?

**Comment by rarkins** [comment]
Use https://docs.renovatebot.com/self-hosted-configuration/#globalextends instead of extends

**Reply by fdw** [reply]
I tried that before and had the same problem :(

**Reply by fdw** [reply]
Do you have any other idea, as both `extends` and `globalextends` produce the same result?

**Reply by rarkins** [reply]
They behave differently, should have a different result. Provide the debug log when you use globalExtends

**Reply by fdw** [reply]
I've let it run once with `extends` and once with `globalExtends` and diffed the debug log. The only changes were regarding times and caching, but `packageFiles with updates` was identical in both runs.

> Provide the debug log when you use globalExtends

Unfortunately, I can't provide a full debug log because of corporate policy. If you can tell me which parts are interesting, I can paste (and censor) them.

**Reply by rarkins** [reply]
To start with you need to provide all debug logs up until `DEBUG: Combined config` for the `globalExtends` approach. `extends` is wrong - no need to test it again.

If it's too difficult to support you in open source, and takes a lot of my time to instruct you, I won't be able to - you should look into commercial support.

**Reply by fdw** [reply]
I think I found my problem after combing through the `trace` log:
We have, as written above, `globalExtends: ["group:monorepos", "group:recommended"]`. This will be resolved and then merged with the rest of our config. However, the resolved groups will be *appended* to our existing config, thus overwriting the `Non-major` package rule. I'm not sure if this is on purpose.

So now I need to find a way to put our `packageRules` at the end. Do you have an idea?

> If it's too difficult to support you in open source, and takes a lot of my time to instruct you, I won't be able to - you should look into commercial support.

Of course, and thanks again for you time. I hope this is now more or less solved :)

**Reply by fdw** [reply]
After thinking more about this, it feels like a bug to me that my explicitly defined `packageRules` are less important than presets. I've looked through the issues (both open and closed), but I couldn't find anything related.

Is this desired behavior? If so, how can I make sure that my `packageRule` is evaluated last?
If not, do you want me to open an issue? If it's not too complicated, I might also try to write a PR.

**Reply by rarkins** [reply]
Please create a minimal reproduction (so logs are not too big) and then provide the logs for me to look at. I expected that globalExtends is resolved _before_ any packageRules

**Reply by fdw** [reply]
That is true, `globalExtends` is resolved before the package rules, but the `packageRules` from `groups:monorepo` are appended to the end of my package rules - so they will be evaluated after my `packageRules` and change things again. I would expect them to be added to the top, so that my explicit configuration "wins".

I retraced that in a `trace` log, but unfortunately, it's gigantic. I think this is the relevant part:
```
TRACE: Post-merge resolve config
       "config": {
         "allowedHeaders": ["X-*"],
         "allowedEnv": [],
         "detectGlobalManagerConfig": false,
         "detectHostRulesFromEnv": false,
         <snip>
         "globalExtends": ["group:monorepos", "group:recommended"],
         "description": [
           "Group known monorepo packages together.",
           "Use curated list of recommended non-monorepo package groupings."
         ],
         "enabled": true,
         <snip>
         "packageRules": [
           {
             "matchPackagePatterns": ["^FluentValidation.*"],
             "groupName": "FluentValidation"
           },
           {
             "matchSourceUrls": ["https://github.com/dotnet/reactive"],
             "groupName": ".NET Reactive extensions"
           },
           {
             "matchDatasources": ["npm"],
             "matchPackagePatterns": ["^@reduxjs/", "react-redux"],
             "groupName": "Reduxjs"
           },
           {"matchUpdateTypes": ["minor", "patch"], "groupName": "Non-major"},
           {
             "description": [
               "accounts monorepo",
               "Group packages from accounts monorepo together."
             ],
             "matchSourceUrls": ["https://github.com/accounts-js/accounts"],
             "groupName": "accounts monorepo",
             "matchUpdateTypes": ["digest", "patch", "minor", "major"]
           },
           {
             "description": [
               "acot monorepo",
               "Group packages from acot monorepo together."
             ],
             "matchSourceUrls": ["https://github.com/acot-a11y/acot"],
             "groupName": "acot monorepo",
             "matchUpdateTypes": ["digest", "patch", "minor", "major"]
           },
           {
             "description": [
               "ag-grid monorepo",
               "Group packages from ag-grid monorepo together."
             ],
             "matchSourceUrls": ["https://github.com/ag-grid/ag-grid"],
             "groupName": "ag-grid monorepo",
             "matchUpdateTypes": ["digest", "patch", "minor", "major"]
           },
           {
             "description": [
               "algolia-instantsearch monorepo",
               "Group packages from algolia-instantsearch monorepo together."
             ],
             "matchSourceUrls": [
               "https://github.com/algolia/instantsearch",
               "https://github.com/algolia/instantsearch.js"
             ],
             "groupName": "algolia-instantsearch monorepo",
             "matchUpdateTypes": ["digest", "patch", "minor", "major"]
           },
           {
             "description": [
               "algolia-react-instantsearch monorepo",
               "Group packages from algolia-react-instantsearch monorepo together."
             ],
             "matchSourceUrls": ["https://github.com/algolia/react-instantsearch"],
             "groupName": "algolia-react-instantsearch monorepo",
             "matchUpdateTypes": ["digest", "patch", "minor", "major"]
           },
           {
             "description": [
               "algoliasearch-autocomplete monorepo",
               "Group packages from algoliasearch-autocomplete monorepo together."
             ],
             "matchSourceUrls": ["https://github.com/algolia/autocomplete"],
             "groupName": "algoliasearch-autocomplete monorepo",
             "matchUpdateTypes": ["digest", "patch", "minor", "major"]
           },
           <snip>
```
Note that the first `packageRules` are mine, especially the `{"matchUpdateTypes": ["minor", "patch"], "groupName": "Non-major"}`. They are followed by the resolved presets.

Do you need more?

**Reply by rarkins** [reply]
I need what I asked for above

**Reply by fdw** [reply]
I guess the reproduction needs to be public? Do you also need access to the runner, or is it enough if I share a log with you afterwards? If so, which log level do you need?

**Reply by rarkins** [reply]
It doesn't in this case need to be public. What I'm after is debug level logs up to the point where you think something is wrong. And for that to be manageable to inspect manually then it needs to be minimal. If this isn't enough, the next step will be to set printConfig=true.

BTW you don't need to run at trace level to see a trace message. If you use logLevelRemap then you can remap selected messages to debug and capture them that way.

**Reply by fdw** [reply]
Alright, I'll see what I can do 🙂

**Reply by fdw** [reply]
So, I have a small reproduction at https://github.com/fdw/renovate-reproduction .
As you can see in the `config.js`, I have a `packageRule` to group `minor` and `patch` level updates. However, there are three other PRs for minor-level updates that have not been grouped as I expected.

The debug log is at https://gist.github.com/fdw/6c63ad84f990d841fa2251b68a45ee3f

Unfortunately, I couldn't make `logLevelRemap` work, but I also have trace.log, which is too large for a gist. Instead, I prettified and uploaded the one line that shows the problem (I think). When merging the `globalExtends`, the `packageRules` are added *after* my `packageRules`. If you need more, please tell me 🙂

**Reply by rarkins** [reply]
Hi, you were right! We made a mistake when implementing globalExtends merge and indeed it gets merged after config.js config instead of before. I will fix it soon - thanks for your patience in helping me understand

**Reply by rarkins** [reply, marked as answer]
Tracking issue: https://github.com/renovatebot/renovate/issues/28131
